### PR TITLE
spring-cloud-reactivefeign-core module added

### LINKED ISSUE #4: Add support for reactive feign clients
See https://github.com/spring-cloud/spring-cloud-netflix/issues/1554 for history.

### PR DESCRIPTION
**Pull request by kptfh**
Fixes #4

### CODE REVIEW

**Comment by kptfh** [comment]
@spencergibb A bit confused with "Some checks were not successful" as I just renamed package and reformatted code

**Comment by ryanjbaxter** [comment]
@kptfh its just because code coverage targets were not hit.  We are not sticklers about it but if you want to try an increase the overage, that is always welcome.

**Comment by kptfh** [comment]
@ryanjbaxter Plan to add some Unit tests but it will take some time.

**Comment by kptfh** [comment]
@spencergibb @ryanjbaxter Renamed package, reformatted code, added tests.

File: spring-cloud-reactivefeign-core/src/main/java/org/springframework/cloud/openfeign/reactive/ReactiveFeign.java [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+import java.lang.reflect.Method;
+import java.lang.reflect.Proxy;
+import java.util.LinkedHashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Function;
+
+import feign.*;
```
Comment (ryanjbaxter) [inline spring-cloud-reactivefeign-core/src/main/java/org/springframework/cloud/openfeign/reactive/ReactiveFeign.java]: We dont use `*` imports

File: spring-cloud-reactivefeign-core/src/main/java/org/springframework/cloud/openfeign/reactive/ReactiveFeign.java [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+		 *
+		 * @param options Feign {@code Request.Options} object
+		 * @return this builder
+		 */
+		public Builder<T> options(final Request.Options options) {
+			boolean tryUseCompression = options instanceof ReactiveOptions
+					&& ((ReactiveOptions) options).isTryUseCompression();
+
+			ReactorClientHttpConnector connector = new ReactorClientHttpConnector(
+					opts -> opts
```
Comment (ryanjbaxter) [inline spring-cloud-reactivefeign-core/src/main/java/org/springframework/cloud/openfeign/reactive/ReactiveFeign.java]: The spacing here makes this code hard to read

Comment (ryanjbaxter) [reply spring-cloud-reactivefeign-core/src/main/java/org/springframework/cloud/openfeign/reactive/ReactiveFeign.java]: Are these timeouts configurable?

File: spring-cloud-reactivefeign-core/src/main/java/org/springframework/cloud/openfeign/reactive/ReactiveFeign.java [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+										TimeUnit.MILLISECONDS));
+							}));
+
+			this.webClient = webClient.mutate().clientConnector(connector).build();
+			return this;
+		}
+
+		public Builder<T> retryWhen(
+				Function<Flux<Throwable>, Publisher<Throwable>> retryFunction) {
+			this.retryFunction = retryFunction;
```
Comment (ryanjbaxter) [inline spring-cloud-reactivefeign-core/src/main/java/org/springframework/cloud/openfeign/reactive/ReactiveFeign.java]: How is the retry configured?

File: spring-cloud-reactivefeign-core/src/main/java/org/springframework/cloud/openfeign/reactive/client/WebReactiveHttpClient.java [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+				.uri(request.uri())
+				.headers(httpHeaders -> request.headers().forEach(
+						(key, value) -> httpHeaders.put(key, (List<String>) value)))
+				.body(bodyActualType != null
+						? BodyInserters.fromPublisher(request.body(), bodyActualType)
+						: BodyInserters.empty())
+				.retrieve()
+				.onStatus(httpStatus -> decode404 && httpStatus == NOT_FOUND,
+						clientResponse -> null)
+				.onStatus(HttpStatus::isError, clientResponse -> clientResponse
```
Comment (ryanjbaxter) [inline spring-cloud-reactivefeign-core/src/main/java/org/springframework/cloud/openfeign/reactive/client/WebReactiveHttpClient.java]: wow, can we break this code up a bit so it is easier to understand?

File: spring-cloud-reactivefeign-core/src/main/java/org/springframework/cloud/openfeign/reactive/ReactiveFeign.java (hunk 1 above)
Comment (kptfh) [reply spring-cloud-reactivefeign-core/src/main/java/org/springframework/cloud/openfeign/reactive/ReactiveFeign.java]: fixed
Do you have formatter settings for Intellij Idea?

File: spring-cloud-reactivefeign-core/src/main/java/org/springframework/cloud/openfeign/reactive/ReactiveFeign.java (hunk 2 above)
Comment (kptfh) [reply spring-cloud-reactivefeign-core/src/main/java/org/springframework/cloud/openfeign/reactive/ReactiveFeign.java]: yes, you can specify them while building Feign client

File: spring-cloud-reactivefeign-core/src/main/java/org/springframework/cloud/openfeign/reactive/ReactiveFeign.java (hunk 3 above)
Comment (kptfh) [reply spring-cloud-reactivefeign-core/src/main/java/org/springframework/cloud/openfeign/reactive/ReactiveFeign.java]: The most generic way to configure retry for Mono or Flux is to provide retry function to retryWhen() method. There is ReactiveRetryers class with predefined retry functions.
This module doesn't have any spring configurations.

File: spring-cloud-reactivefeign-core/src/main/java/org/springframework/cloud/openfeign/reactive/ReactiveFeign.java (hunk 1 above)
Comment (vasilievip) [reply spring-cloud-reactivefeign-core/src/main/java/org/springframework/cloud/openfeign/reactive/ReactiveFeign.java]: @kptfh  https://plugins.jetbrains.com/plugin/6546-eclipse-code-formatter

File: spring-cloud-reactivefeign-core/src/main/java/org/springframework/cloud/openfeign/reactive/client/WebReactiveHttpClient.java (hunk 4 above)
Comment (kptfh) [reply spring-cloud-reactivefeign-core/src/main/java/org/springframework/cloud/openfeign/reactive/client/WebReactiveHttpClient.java]: fixed

File: spring-cloud-reactivefeign-core/src/main/java/org/springframework/cloud/openfeign/reactive/ReactiveFeign.java (hunk 1 above)
Comment (kptfh) [reply spring-cloud-reactivefeign-core/src/main/java/org/springframework/cloud/openfeign/reactive/ReactiveFeign.java]: @vasilievip didn't work as expected. Actually this code is the result of applying this formmater with those settings.

File: spring-cloud-reactivefeign-core/src/main/java/org/springframework/cloud/openfeign/reactive/ReactiveFeign.java (hunk 3 above)
Comment (kptfh) [reply spring-cloud-reactivefeign-core/src/main/java/org/springframework/cloud/openfeign/reactive/ReactiveFeign.java]: This module is the most generic code that may be used without any cloud specific code and spring configurations. In cloud related module retries will be configured via Ribbon. Also spring configuration module is planned to be added later.

File: spring-cloud-reactivefeign-core/src/main/java/org/springframework/cloud/openfeign/reactive/ReactiveFeign.java (hunk 1 above)
Comment (ryanjbaxter) [reply spring-cloud-reactivefeign-core/src/main/java/org/springframework/cloud/openfeign/reactive/ReactiveFeign.java]: See https://github.com/spring-cloud/spring-cloud-openfeign/blob/master/.github/CONTRIBUTING.md#code-conventions-and-housekeeping

File: spring-cloud-reactivefeign-core/src/main/java/org/springframework/cloud/openfeign/reactive/ReactiveFeign.java (hunk 3 above)
Comment (ryanjbaxter) [reply spring-cloud-reactivefeign-core/src/main/java/org/springframework/cloud/openfeign/reactive/ReactiveFeign.java]: So we can continue to use `client.ribbon.MaxAutoRetries`, `client.ribbon.MaxAutoRetriesNextServer`, and `client.ribbon.OkToRetryOnAllOperations`?

Comment (kptfh) [reply spring-cloud-reactivefeign-core/src/main/java/org/springframework/cloud/openfeign/reactive/ReactiveFeign.java]: Yes. After some work on configuration will be done. If you check [RibbonReactiveClient](https://github.com/kptfh/feign-reactive/blob/master/feign-reactive-cloud/src/main/java/feign/reactive/client/RibbonReactiveClient.java) you will see that it is based on com.netflix.loadbalancer.reactive.LoadBalancerCommand and it's configured via regular com.netflix.client.RetryHandler.

**Comment by ryanjbaxter** [comment]
We need to add some documentation.  Also I think you need to add the module to the starter pom.

**Comment by riba2101** [comment]
Hey guys, 

if you were building/testing something here's a workaround for the time being:

`public class ReactiveDecoder implements Decoder {

    private final Decoder delegate;

    public ReactiveDecoder(Decoder delegate) {
        Objects.requireNonNull(delegate, "Decoder must not be null. ");
        this.delegate = delegate;
    }

    @Override
    public Object decode(Response response, Type type) throws IOException {
        boolean mono = isMono(type);
        boolean flux = isFlux(type);

        if (!flux && !mono) {
            return delegate.decode(response, type);
        }

        if (response.status() == 404) {
            if (mono) {
                return Mono.empty();
            }

            return Flux.empty();
        }

        if (mono) {
            Type enclosedType = Util.resolveLastTypeParameter(type, Mono.class);
            return Mono.just(delegate.decode(response, enclosedType));
        }

        Type enclosedType = Util.resolveLastTypeParameter(type, Flux.class);
        return Flux.just(delegate.decode(response, enclosedType));
    }

    private boolean isFlux(Type type) {
        if (!(type instanceof ParameterizedType)) {
            return false;
        }
        ParameterizedType parameterizedType = (ParameterizedType) type;
        return parameterizedType.getRawType().equals(Flux.class);
    }

    private boolean isMono(Type type) {
        if (!(type instanceof ParameterizedType)) {
            return false;
        }
        ParameterizedType parameterizedType = (ParameterizedType) type;
        return parameterizedType.getRawType().equals(Mono.class);
    }

}`

and then substitute the default bean for this one:

`    @Autowired
    private ObjectFactory<HttpMessageConverters> messageConverters;

    @Bean
    public Decoder feignDecoder() {
        return new ReactiveDecoder(new OptionalDecoder(new ResponseEntityDecoder(new SpringDecoder(this.messageConverters))));
    }`

this worked for me.

Hope it helps someone

**Comment by ryanjbaxter** [comment]
@kptfh do you plan on picking this back up?

**Comment by kptfh** [comment]
@ryanjbaxter Sorry, but your questions is not clear to me. Do you mean take a look at @riba2101's comment?

**Comment by kptfh** [comment]
Sorry treated your inactivity as you've lost any interest to this PR. Seemed that I fixed all your comments and had no feedback. So I made some bunch of changes in original repository and will need to redo all this stuff with renaming of packages again. Actually I still have a strong intention to complete this work but will not be able to do it without your help and good communication.

**Comment by ryanjbaxter** [comment]
I was waiting for you to add documentation as I asked for above.

**Comment by kptfh** [comment]
Literally, you said: 
>  We need to add some documentation. Also I think you need to add the module to the starter pom
So I added module to the starter pom but didn't expect that you mean me to add doc.
Finally it's not clear what kind of doc to add: java doc(which classes?), how-to doc?

**Comment by ryanjbaxter** [comment]
We would want documentation on how to use the feature added to https://github.com/spring-cloud/spring-cloud-openfeign/blob/master/docs/src/main/asciidoc/spring-cloud-openfeign.adoc

**Comment by kptfh** [comment]
Hi @ryanjbaxter. Added adoc to spring-cloud-reactivefeign-core. Want to remind that this module does't support Ribbon or Hystrix. This support will be added in the next "spring-cloud-reactivefeign-cloud" module

**Comment by kptfh** [comment]
Also faced some issue with Eclipse formatter with Spring Cloud settings. You may note that some method calls are not aligned properly.

**Comment by ryanjbaxter** [comment]
> This support will be added in the next "spring-cloud-reactivefeign-cloud" module

what does this mean?

**Comment by kptfh** [comment]
It was agreed with @spencergibb  that there are 2 modules:
spring-cloud-reactivefeign-core -- core logic
spring-cloud-reactivefeign-cloud -- cloud specific logic (ribbon/hystrix support)
To keep pull requests small decided to include in this PR only core module.
Here you can see these modules [https://github.com/kptfh/feign-reactive](https://github.com/kptfh/feign-reactive)

**Comment by ryanjbaxter** [comment]
Got it.

I dont think the docs for this need to be its own file, it is fine to add them to the existing docs file.

**Comment by kptfh** [comment]
@ryanjbaxter appended to spring-cloud-openfeign.adoc

**Comment by kptfh** [comment]
Hi, @ryanjbaxter. Any comments or proposals?

**Comment by ryanjbaxter** [comment]
Thanks @kptfh.  Right now we are going to focus on getting the Finchley RELEASE out the door.  I am going to label this for SR1 for now.  It might take me a few days to get to look at this again.

**Comment by kptfh** [comment]
@ryanjbaxter Does it make sense to add spring-cloud-reactivefeign-cloud module (cloud specific logic - ribbon/hystrix support ) to this pull request to review it all together?

**Comment by vasilievip** [comment]
@ryanjbaxter relase is out, can you merge changes or there is additional work required?

**Comment by ryanjbaxter** [comment]
Hopefully we can get this merged this week.  We might be doing another Finchley service release earlier than planned do to an issue that popped up in another project.  So I might wait until after that.

File: spring-cloud-reactivefeign-core/src/main/java/org/springframework/cloud/openfeign/reactive/DefaultMethodHandler.java [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+import java.lang.reflect.Field;
+import java.lang.reflect.Method;
+
+import feign.InvocationHandlerFactory.MethodHandler;
+
+/**
+ * Handles default methods by directly invoking the default method code on the interface.
+ * The bindTo method must be called on the result before invoke is called.
+ */
+final class DefaultMethodHandler implements MethodHandler {
```
Comment (kdavisk6) [inline spring-cloud-reactivefeign-core/src/main/java/org/springframework/cloud/openfeign/reactive/DefaultMethodHandler.java]: @spencergibb @ryanjbaxter This class is a direct copy from the `feign` core.  I don't think you'll want to accept this verbatim.  This class is not exposed in the public `feign` api.

File: spring-cloud-reactivefeign-core/src/main/java/org/springframework/cloud/openfeign/reactive/ReactiveFeign.java (hunk 3 above)
Comment (kdavisk6) [reply spring-cloud-reactivefeign-core/src/main/java/org/springframework/cloud/openfeign/reactive/ReactiveFeign.java]: Coming in a little late here, but why not contribute this `Feign` core?  @kptfh you've lifted almost the entire class from `feign-core`?

**Comment by kdavisk6** [comment]
I am coming into this very late, and I don't want to derail this, but coming from `feign`, this PR contains more `feign` code than `Spring`.  In fact, with a little more refactoring, this could be a PR into `feign` directly.  I'm confused as to why @kptfh you did not consider submitting this to `feign`?

**Comment by spencergibb** [comment]
I hadn't realized that with a minor refactor it could go in an open feign module. I think that might be a good idea. Reactor could certainly be used outside of spring.

**Comment by kdavisk6** [comment]
I don't see any reason why we could not create a `feign-reactive-reactor` sub module with the changes included in `spring-cloud-reactivefeign-core` module.  The refactoring would be around the `Retry`, as the current `Retryer` interface in `feign` would not work in this case, but I'm sure we can make it work.

This feature has been widely requested within feign as  mentioned in issues https://github.com/OpenFeign/feign/issues/622 and https://github.com/OpenFeign/feign/issues/644 and https://github.com/OpenFeign/feign/issues/361.  `feign` is now baselined on JDK 8, so I don't see anything really in the way here.

I'm more than willing to help out where I can with this.

**Comment by kptfh** [comment]
@kdavisk6 This version of reactive feign depends on Spring's WebClient as reactive http transport.
Also don't see a lot of benefits to use reactive approach on Android as it suits more highly loaded server side (saves memory and CPU on reducing number of threads)

**Comment by kdavisk6** [comment]
Looking over the PR again, I can see that the `WebReactiveHttpClient` is the only component that relies on Spring.  The rest of the PR is pure Feign.

The way this PR is constructed currently, it copies a few key core objects from `feign-core` verbatim into `spring-cloud`.  This could cause issues in the future.  For example, #34 requests that the `spring-cloud` update to `OpenFeign` 10.x, which will break this integration.  Also, @spencergibb has mentioned in the #4 that he would prefer that nothing of `feign` make it into `spring-boot`.  This PR will do just that, it will now have it's own *copy* of `feign` will be decoupled from `feign` without a fork.

One way to ensure that both projects can benefit is to take the Feign related items such as `ReactiveHttpClient`, `ReactiveFeign`, `ReactiveEncoder`, `ReactiveDecoder`, etc... and add them to `OpenFeign` under `feign-reactive` and keep the `WebReactiveHttpClient` here with Spring Cloud.  This way, both projects get the benefit.  `feign` will have the reactive framework and Spring is still able to take advantage of it without the duplication.  Thoughts?

**Comment by kdavisk6** [comment]
@kptfh We can certainly create a repository under OpenFeign similar to what was done for Vert.x for this work.  I'm volunteering my time to help you make this happen.  To help coordinate, can we move that discussion over to https://github.com/OpenFeign/feign/issues/644?

**Comment by kptfh** [comment]
@spencergibb  what do you think about this?
1) Leave PR here as is or 
2) split into open feign part (reactive framework ReactiveHttpClient  without implementation)  and Spring part (based on WebClient)?

**Comment by ryanjbaxter** [comment]
I think it makes the most sense to keep as much as we can in OpenFeign.

**Comment by spencergibb** [comment]
:+1:

**Comment by kptfh** [comment]
@spencergibb @ryanjbaxter Can you comment on this issue https://github.com/OpenFeign/feign-reactive/issues/15  ?

**Comment by kptfh** [comment]
@ryanjbaxter @spencergibb stack into dead end with openfeign repo. Don't see any benefits to keep code there. No one is reviewing the code except one guy (@kdavisk6) that don't understand how reactive approach should work. It will last for years and from my side looks like sabotage or time killing.
Please, take a look at this issue [https://github.com/OpenFeign/feign-reactive/issues/15](https://github.com/OpenFeign/feign-reactive/issues/15) and pull request [https://github.com/OpenFeign/feign-reactive/pull/18](https://github.com/OpenFeign/feign-reactive/pull/18)

**Comment by Aloren** [comment]
hello @spencergibb, could you please advise what should be done to proceed communication on this pr? obviously, there is no common vision of what should or should not be done with reactive-feign right now. maybe we could find some temporary solutions for the not agreed issues arisen during review?

**Comment by ryanjbaxter** [comment]
@Aloren we do not want to duplicate code from openfeign in this repo.  I would rather keep this downstream.  From https://github.com/OpenFeign/feign-reactive/issues/15 it sounds like the downstream code was going to be moved to https://github.com/kptfh/feign-reactive.  Maybe the best approach then is to have a separate corresponding repo for an reactive spring cloud openfeign module that depends on that code.  I could talk to the team and see if they would mind that being a community driven project in our incubator if the community is interested in that.  Once the code in https://github.com/kptfh/feign-reactive can be moved into OpenFeign then we consider moving that incubator module into this project.

**Comment by OlgaMaciaszek** [comment]
Closing, as the team has taken the decision that we will add integration only whenever there's reactive support in OpenFeign core.

**Pull request closed without merging**